# Rocq upstream fixes found during the Lean port (apply to master with `git apply`).
# UNTESTED: no Rocq toolchain on this machine. (1) bind package Assumptions in
# wp axioms that otherwise hold for every go.Semantics; (2) store_array reads
# index j, not n.

diff --git a/new/golang/defn/array.v b/new/golang/defn/array.v
index 65a53e9da..4c46188bc 100644
--- a/new/golang/defn/array.v
+++ b/new/golang/defn/array.v
@@ -42,7 +42,7 @@ Class ArraySemantics `{!GoSemanticsFunctions} :=
     (foldl (λ str_so_far j,
                 str_so_far;;
                 let elem_addr := IndexRef (go.ArrayType n elem_type) (l, #(W64 j)) in
-                let elem_val := Index (go.ArrayType n elem_type) (v, #(W64 n)) in
+                let elem_val := Index (go.ArrayType n elem_type) (v, #(W64 j)) in
                 GoStore elem_type (elem_addr, elem_val))
              (#()) (seqZ 0 n)
     )%E;
diff --git a/new/proof/go_etcd_io/etcd/api/v3/etcdserverpb.v b/new/proof/go_etcd_io/etcd/api/v3/etcdserverpb.v
index f7e2c5318..121f8a773 100644
--- a/new/proof/go_etcd_io/etcd/api/v3/etcdserverpb.v
+++ b/new/proof/go_etcd_io/etcd/api/v3/etcdserverpb.v
@@ -27,7 +27,7 @@ Axiom own_InternalRaftRequest_new_header :
   ∃ req_abs',
     own_InternalRaftRequest (req <| etcdserverpb.InternalRaftRequest.Header' := hdr_ptr |>) req_abs'.
 
-Axiom wp_InternalRaftRequest__Marshal : ∀ m_ptr m msg,
+Axiom wp_InternalRaftRequest__Marshal : ∀ `{!etcdserverpb.Assumptions} m_ptr m msg,
   {{{ is_pkg_init etcdserverpb ∗ m_ptr ↦ m ∗ own_InternalRaftRequest m msg }}}
     m_ptr @! (go.PointerType etcdserverpb.InternalRaftRequest) @! "Marshal" #()
   {{{ dAtA_sl (err : error.t), RET (#dAtA_sl, #err);
diff --git a/new/proof/go_etcd_io/etcd/cache/v3.v b/new/proof/go_etcd_io/etcd/cache/v3.v
index 38049c998..6ef884dcc 100644
--- a/new/proof/go_etcd_io/etcd/cache/v3.v
+++ b/new/proof/go_etcd_io/etcd/cache/v3.v
@@ -48,7 +48,7 @@ Context [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] [V] `[!IntoValTyped T'
     {is_item : T' → V → iProp Σ} {rev_item : V → w64}.
 
 Axiom wp_ringBuffer__PeekOldest :
-  ∀ r buf,
+  ∀ `{!cache.Assumptions} r buf,
   {{{ is_pkg_init cache ∗
       own_ringBuffer r is_item rev_item buf }}}
     r @! (go.PointerType (cache.ringBuffer T)) @! "PeekOldest" #()
@@ -58,7 +58,7 @@ Axiom wp_ringBuffer__PeekOldest :
 (* [P i] is the invariant that holds after [iter] has been called [i] times on
    the appropriate items. *)
 Axiom wp_ringBuffer__DescendLessOrEqual :
-  ∀ P (pivot : w64) (iter : func.t) r buf Φ,
+  ∀ `{!cache.Assumptions} P (pivot : w64) (iter : func.t) r buf Φ,
   let iter_itemvs := reverse (filter (λ item, sint.Z (rev_item item) ≤ sint.Z pivot) buf) in
   is_pkg_init cache ∗
   own_ringBuffer r is_item rev_item buf ∗
diff --git a/new/proof/go_etcd_io/etcd/client/v3_proof/client.v b/new/proof/go_etcd_io/etcd/client/v3_proof/client.v
index 835ae5e76..4bbaa4d93 100644
--- a/new/proof/go_etcd_io/etcd/client/v3_proof/client.v
+++ b/new/proof/go_etcd_io/etcd/client/v3_proof/client.v
@@ -33,7 +33,7 @@ Global Existing Instance is_Client_pers.
 Axiom N : namespace.
 
 (* Only specifying Do Get for now. *)
-Axiom wp_Client__Do_Get : ∀ key client γ (ctx : context.Context.t) (op : clientv3.Op.t),
+Axiom wp_Client__Do_Get : ∀ `{!clientv3.Assumptions} key client γ (ctx : context.Context.t) (op : clientv3.Op.t),
   ∀ Φ,
   (is_Client client γ ∗
    is_Op op (Op.Get (RangeRequest.default <| RangeRequest.key := key |>))) -∗
@@ -44,19 +44,19 @@ Axiom wp_Client__Do_Get : ∀ key client γ (ctx : context.Context.t) (op : clie
   WP client @! clientv3.Client @! "Do" #ctx #op {{ Φ }}.
 
 Axiom wp_Client__GetLogger :
-  ∀ (client : loc) γ,
+  ∀ `{!clientv3.Assumptions} (client : loc) γ,
   {{{ is_Client client γ }}}
     client @! (go.PointerType clientv3.Client) @! "GetLogger" #()
   {{{ (lg : loc), RET #lg; True }}}.
 
 Axiom wp_Client__Ctx :
-  ∀ (client : loc) γ,
+  ∀ `{!clientv3.Assumptions} (client : loc) γ,
   {{{ is_Client client γ }}}
     client @! (go.PointerType clientv3.Client) @! "Ctx" #()
   {{{ ctx s, RET #(interface.ok ctx); is_Context ctx s }}}.
 
 Axiom wp_Client__Grant :
-  ∀ client γ (ctx : context.Context.t) (ttl : w64),
+  ∀ `{!clientv3.Assumptions} client γ (ctx : context.Context.t) (ttl : w64),
   {{{ is_Client client γ }}}
     client @! (go.PointerType clientv3.Client) @! "Grant" #ctx #ttl
   {{{
@@ -69,7 +69,7 @@ Axiom wp_Client__Grant :
   }}}.
 
 Axiom wp_Client__KeepAlive :
-  ∀ client γ ctx id,
+  ∀ `{!clientv3.Assumptions} client γ ctx id,
   (* The precondition requires that this is only called on a `Grant`ed lease. *)
   {{{ is_Client client γ ∗ is_etcd_lease γ id }}}
     client @! (go.PointerType clientv3.Client) @! "KeepAlive" #(interface.ok ctx) #id
diff --git a/new/proof/go_etcd_io/etcd/server/v3/etcdserver.v b/new/proof/go_etcd_io/etcd/server/v3/etcdserver.v
index 5c6dedfd3..1367072f9 100644
--- a/new/proof/go_etcd_io/etcd/server/v3/etcdserver.v
+++ b/new/proof/go_etcd_io/etcd/server/v3/etcdserver.v
@@ -88,12 +88,12 @@ Existing Instance is_EtcdServer_internal_pers.
    `own_EtcdServer`.
  *)
 
-Axiom wp_EtcdServer__getAppliedIndex : ∀ s γ,
+Axiom wp_EtcdServer__getAppliedIndex : ∀ `{!etcdserver.Assumptions} s γ,
   {{{ is_pkg_init etcdserver ∗ is_EtcdServer_internal s γ }}}
     s @! (go.PointerType etcdserver.EtcdServer) @! "getAppliedIndex" #()
   {{{ (a : w64), RET #a; True }}}.
 
-Axiom wp_EtcdServer__getCommittedIndex : ∀ s γ,
+Axiom wp_EtcdServer__getCommittedIndex : ∀ `{!etcdserver.Assumptions} s γ,
   {{{ is_pkg_init etcdserver ∗ is_EtcdServer_internal s γ }}}
     s @! (go.PointerType etcdserver.EtcdServer) @! "getCommittedIndex" #()
   {{{ (a : w64), RET #a; True }}}.
@@ -104,7 +104,7 @@ Definition is_SimpleRequest r : iProp Σ :=
   "_" ∷ True
 .
 
-Axiom wp_EtcdServer__AuthInfoFromCtx : ∀ s γ ctx ctx_desc,
+Axiom wp_EtcdServer__AuthInfoFromCtx : ∀ `{!etcdserver.Assumptions} s γ ctx ctx_desc,
   {{{ is_pkg_init etcdserver ∗ own_EtcdServer s γ ∗ is_Context ctx ctx_desc }}}
     s @! (go.PointerType etcdserver.EtcdServer) @! "AuthInfoFromCtx" #(interface.ok ctx)
   {{{ (a_ptr : loc) (err : interface.t), RET (#a_ptr, #err);
diff --git a/new/proof/go_etcd_io/raft/v3_proof/readonly.v b/new/proof/go_etcd_io/raft/v3_proof/readonly.v
index b0e65b01b..58b6c1f26 100644
--- a/new/proof/go_etcd_io/raft/v3_proof/readonly.v
+++ b/new/proof/go_etcd_io/raft/v3_proof/readonly.v
@@ -716,7 +716,7 @@ Definition own_AckedIndexer (i : interface.t_ok) (acks : gmap w64 w64) I : iProp
         I }}}).
 
 Axiom wp_JointConfig__CommittedIndex :
-  ∀ l acks (c : quorum.JointConfig.t) voters_ref (voters : gmap w64 ()) I,
+  ∀ `{!quorum.Assumptions} l acks (c : quorum.JointConfig.t) voters_ref (voters : gmap w64 ()) I,
   {{{
       is_pkg_init quorum ∗
       "Hl" ∷ own_AckedIndexer l acks I ∗
diff --git a/new/proof/k8s_io/utils/third_party/forked/golang/btree.v b/new/proof/k8s_io/utils/third_party/forked/golang/btree.v
index da2d3e1da..71ead93d3 100644
--- a/new/proof/k8s_io/utils/third_party/forked/golang/btree.v
+++ b/new/proof/k8s_io/utils/third_party/forked/golang/btree.v
@@ -27,7 +27,7 @@ Axiom own_BTree_dfractional :
 #[global] Existing Instance own_BTree_dfractional.
 
 Axiom wp_BTree__Clone :
-  ∀ [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
+  ∀ `{!btree.Assumptions} [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
     (is_item : ∀ (p : T') (v : V), iProp Σ) (less : V → V → Prop)
     (items : list V),
 
@@ -40,7 +40,7 @@ Axiom wp_BTree__Clone :
       own_BTree t' is_item less items (DfracOwn 1) }}}.
 
 Axiom wp_BTree__Get :
-  ∀ [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
+  ∀ `{!btree.Assumptions} [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
     (is_item : ∀ (p : T') (v : V), iProp Σ) (less : V → V → Prop)
     (items : list V),
 
@@ -57,7 +57,7 @@ Axiom wp_BTree__Get :
   }}}.
 
 Axiom wp_BTree__ReplaceOrInsert :
-  ∀ [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
+  ∀ `{!btree.Assumptions} [T'] `[!ZeroVal T'] `[!TypedPointsto (Σ:=Σ) T'] `[!IntoValTyped T' T] [V]
     (is_item : ∀ (p : T') (v : V), iProp Σ) (less : V → V → Prop)
     (items : list V),
 
diff --git a/new/proof/time.v b/new/proof/time.v
index 8d39a1d93..314a39a8c 100644
--- a/new/proof/time.v
+++ b/new/proof/time.v
@@ -71,16 +71,17 @@ Proof.
 Qed.
 
 Axiom wp_Now :
+  ∀ `{!time.Assumptions},
   {{{ True }}}
     @! time.Now #()
   {{{ (t : time.Time.t), RET #t; True }}}.
 
-Axiom wp_Until : ∀ (deadline : time.Time.t) ,
+Axiom wp_Until : ∀ `{!time.Assumptions} (deadline : time.Time.t) ,
   {{{ True }}}
     @! time.Until #deadline
   {{{ (x : w64), RET #x; True }}}.
 
-Axiom wp_Time__Add : ∀ (t : time.Time.t) (d : time.Duration.t),
+Axiom wp_Time__Add : ∀ `{!time.Assumptions} (t : time.Time.t) (d : time.Duration.t),
   {{{ True }}}
     t @! time.Time @! "Add" #d
   {{{ (t : time.Time.t), RET #t; True }}}.
